File: solutions/go/0068_文本左右对齐.go
package main

import (
	"fmt"
	"strings"
)

func fullJustify(words []string, maxWidth int) []string {
	var result []string
	i := 0

	for i < len(words) {
		lineLen := len(words[i])
		j := i + 1
		for j < len(words) && lineLen+1+len(words[j]) <= maxWidth {
			lineLen += 1 + len(words[j])
			j++
		}

		lineWords := words[i:j]
		totalChars := 0
		for _, w := range lineWords {
			totalChars += len(w)
		}
		gaps := len(lineWords) - 1
		isLastLine := j == len(words)

		if gaps == 0 || isLastLine {
			line := strings.Join(lineWords, " ")
			line += strings.Repeat(" ", maxWidth-len(line))
			result = append(result, line)
		} else {
			totalSpaces := maxWidth - totalChars
			spacePerGap := totalSpaces / gaps
			extraSpaces := totalSpaces % gaps
			var line strings.Builder
			for k := 0; k < len(lineWords); k++ {
				line.WriteString(lineWords[k])
				if k < gaps {
					spaces := spacePerGap
					if extraSpaces > 0 {
						spaces++
						extraSpaces--
					}
					line.WriteString(strings.Repeat(" ", spaces))
				}
			}
			result = append(result, line.String())
		}

		i = j
	}

	return result
}

func main() {
	words1 := []string{"This", "is", "an", "example", "of", "text", "justification."}
	result1 := fullJustify(words1, 16)
	for _, line := range result1 {
		fmt.Printf("\"%s\"\n", line)
	}

	words2 := []string{"What", "must", "be", "acknowledgment", "shall", "be"}
	result2 := fullJustify(words2, 16)
	for _, line := range result2 {
		fmt.Printf("\"%s\"\n", line)
	}

	words3 := []string{"Science", "is", "what", "we", "understand", "well", "enough", "to", "explain", "to", "a", "computer.", "Art", "is", "everything", "else", "we", "do"}
	result3 := fullJustify(words3, 20)
	for _, line := range result3 {
		fmt.Printf("\"%s\"\n", line)
	}
}

File: solutions/java/0068_文本左右对齐.java
package com.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new ArrayList<>();
        int i = 0;

        while (i < words.length) {
            int lineLen = words[i].length();
            int j = i + 1;
            while (j < words.length && lineLen + 1 + words[j].length() <= maxWidth) {
                lineLen += 1 + words[j].length();
                j++;
            }

            String[] lineWords = new String[j - i];
            for (int k = 0; k < j - i; k++) {
                lineWords[k] = words[i + k];
            }

            int totalChars = 0;
            for (String w : lineWords) {
                totalChars += w.length();
            }
            int gaps = lineWords.length - 1;
            boolean isLastLine = j == words.length;

            if (gaps == 0 || isLastLine) {
                StringBuilder sb = new StringBuilder(String.join(" ", lineWords));
                while (sb.length() < maxWidth) {
                    sb.append(' ');
                }
                result.add(sb.toString());
            } else {
                int totalSpaces = maxWidth - totalChars;
                int spacePerGap = totalSpaces / gaps;
                int extraSpaces = totalSpaces % gaps;
                StringBuilder sb = new StringBuilder();
                for (int k = 0; k < lineWords.length; k++) {
                    sb.append(lineWords[k]);
                    if (k < gaps) {
                        int spaces = spacePerGap + (extraSpaces > 0 ? 1 : 0);
                        if (extraSpaces > 0) {
                            extraSpaces--;
                        }
                        for (int s = 0; s < spaces; s++) {
                            sb.append(' ');
                        }
                    }
                }
                result.add(sb.toString());
            }

            i = j;
        }

        return result;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        String[] words1 = {"This", "is", "an", "example", "of", "text", "justification."};
        List<String> result1 = sol.fullJustify(words1, 16);
        for (String line : result1) {
            System.out.println("\"" + line + "\"");
        }

        String[] words2 = {"What", "must", "be", "acknowledgment", "shall", "be"};
        List<String> result2 = sol.fullJustify(words2, 16);
        for (String line : result2) {
            System.out.println("\"" + line + "\"");
        }

        String[] words3 = {"Science", "is", "what", "we", "understand", "well", "enough", "to", "explain", "to", "a", "computer.", "Art", "is", "everything", "else", "we", "do"};
        List<String> result3 = sol.fullJustify(words3, 20);
        for (String line : result3) {
            System.out.println("\"" + line + "\"");
        }
    }
}
